File: 2685.count-the-number-of-complete-components.go
package leetcode

import "fmt"

/*
 * @lc app=leetcode id=2685 lang=golang
 *
 * [2685] Count the Number of Complete Components
 */

// @lc code=start
func countCompleteComponents(n int, edges [][]int) int {
	edgeMap := map[int][]int{} 
	for _, e := range edges{
		edgeMap[e[0]] = append(edgeMap[e[0]], e[1])
		edgeMap[e[1]] = append(edgeMap[e[1]], e[0])
	}
	count := 0
	visited := make([]bool, n)//Have we visited[node] before?
	//The idea is to DFS thorugh each subgraph and count the edges and nodes within
	for start := 0; start < n; start++{//outer loop to makes sure we process all subgraphs 
		if visited[start]{
			continue
		}
		//start dfs from start node 
		visited[start] = true
		edgeCount := 0//Actually counts directed edges, so it will be doubled
		nodeCount := 0
		stack := []int{start}
		for len(stack)> 0{
			curr := stack[len(stack)-1]//Pop from the stack
			stack = stack[:len(stack)-1]

			nodeCount++
			edgeCount += len(edgeMap[curr])//This should return empyt slices for nodes with no edges
			for _, next := range edgeMap[curr]{
				if !visited[next]{
					stack = append(stack, next)
					visited[next] = true//Mark next as visited
				}
			}
		}
		//A complete component most have 2*|E| = |V| * (|V|-1)
		if edgeCount== nodeCount *(nodeCount -1){
			count++
		}
	}
	return count
}
// @lc code=end

File: 2685.count-the-number-of-complete-components.py
#
# @lc app=leetcode id=2685 lang=python3
#
# [2685] Count the Number of Complete Components
#

# @lc code=start
from collections import defaultdict
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        """
        start dfs from each node
        during dfs. 
        count the number of edges on the first node
        all following nodes we visit must have the same number of edges
        and it hace to be m-1 of them
        """

        visited = [False] * n
        graph = defaultdict(list)

        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        result = 0
        for i in range(n):
            if visited[i]:
                #this nodes belongs to a already processed component
                continue
            
            visited[i] = True
            component = []
            stack = [i]

            while stack:
                curr = stack.pop()
                component.append(curr)

                for neigh in graph[curr]:
                    if not visited[neigh]:
                        visited[neigh] = True
                        stack.append(neigh)
            
            num_nodes = len(component)
            is_complete = True
            for node in component:
                if len(graph[node]) != num_nodes -1:
                    is_complete = False
                    break

            result += is_complete
        return result

            
            
        
        
# @lc code=end

